contracts/SimpleQuests.sol

```diff
@@ -0,0 +1,416 @@
+        _;
+    }
+
+    // FUNCTIONS
+
+    function generateRequestQuestSeed(uint256 characterID) assertQuestsEnabled assertOwnsCharacter(characterID) public {
+        safeRandoms.requestSingleSeed(msg.sender, RandomUtil.combineSeeds(SEED_RANDOM_QUEST, characterID));
+    }
+
+    function requestQuest(uint256 characterID) public assertQuestsEnabled assertNotOnQuest(characterID) returns (uint256) {
+        return assignNewQuest(characterID);
+    }
+
+    function assignNewQuest(uint256 characterID) private returns (uint256) {
+        uint256 seed = safeRandoms.popSingleSeed(msg.sender, RandomUtil.combineSeeds(SEED_RANDOM_QUEST, characterID), true, true);
+        uint256 currentReputation = characters.getNftVar(characterID, characters.NFTVAR_REPUTATION());
+        uint256 reputationTier;
+        if (currentReputation > vars[VAR_REPUTATION_LEVEL_5]) {
+            reputationTier = 4;
+        } else if (currentReputation > vars[VAR_REPUTATION_LEVEL_4]) {
+            reputationTier = 3;
+        } else if (currentReputation > vars[VAR_REPUTATION_LEVEL_3]) {
+            reputationTier = 2;
+        } else if (currentReputation > vars[VAR_REPUTATION_LEVEL_2]) {
+            reputationTier = 1;
+        } else {
+            reputationTier = 0;
+        }
+        uint256[4] memory chances = tierChances[reputationTier];
+        uint256 tierRoll = RandomUtil.randomSeededMinMax(1, 100, seed);
+        seed = RandomUtil.combineSeeds(seed, 1);
+        Quest memory quest;
+        if (tierRoll > chances[3]) {
+            quest = generateNewQuest(vars[VAR_LEGENDARY_TIER], seed);
+        }
+        else if (tierRoll > chances[2]) {
+            quest = generateNewQuest(vars[VAR_EPIC_TIER], seed);
+        }
+        else if (tierRoll > chances[1]) {
+            quest = generateNewQuest(vars[VAR_RARE_TIER], seed);
+        }
+        else if (tierRoll > chances[0]) {
+            quest = generateNewQuest(vars[VAR_UNCOMMON_TIER], seed);
+        }
+        else {
+            quest = generateNewQuest(vars[VAR_COMMON_TIER], seed);
+        }
+        characterQuest[characterID] = quest.id;
+        uint256[] memory fields = new uint256[](2);
+        fields[0] = characters.NFTVAR_SIMPLEQUEST_PROGRESS();
+        fields[1] = characters.NFTVAR_SIMPLEQUEST_TYPE();
+        uint256[] memory values = new uint256[](2);
+        values[0] = 0;
+        values[1] = uint256(quest.requirementType);
+        characters.setNFTVars(characterID, fields, values);
+        emit QuestAssigned(quest.id, characterID);
+        return quest.id;
+    }
+
+    function generateNewQuest(uint256 tier, uint256 seed) private returns (Quest memory) {
+        Quest[] memory tierQuestTemplates = questTemplates[tier];
+        uint256 index = RandomUtil.randomSeededMinMax(0, tierQuestTemplates.length - 1, seed);
+        Quest memory quest = tierQuestTemplates[index];
+        quest.id = nextQuestID++;
+        quests[quest.id] = quest;
```

Comment:
this little maneuver here costs 150k gas I think. (just this one line, not including the rest of the tx which could be ~281-296k as-is)
Basically, you are writing the entire quest struct worth of data for each time a player gets a new quest. That's $0.285 on bsc at this time.
I only just noticed the template system during this review. In my mind, quest IDs would be essentially a quest template index. We'd store quest IDs in the tiered arrays instead of arrays of quest structs and the quests mapping would be where all the templates are held.

Of course this means if we change a quest, it will affect players who are currently on that quest. The current approach prevents template changes from affecting players who are already on a quest, but the reference way would allow us to plug something damaging.

I was envisioning a scenario where many many templates would be created, not all of them necessarily stored in the tiered arrays (so we could add or remove quests from availability as we see fit), and ideally we'd never edit them unless necessary for an emergency. That way we could save considerable gas for the user without smaller tweaks affecting them (we'd just create a new, similar quest instead and swap the reference in the tiered arrays)

I understand this could be a pretty big change so please reach out in DMs so we can discuss it

Reply:
Hello, random stranger here

I noticed that you tend to put specific gas estimations. How do you come up with these numbers? Is there a test you run?

Regarding the quest mapping. Are you saying that have an array of "quests" where each index represents a position for the quest template inside another variable, a map where all quest templates are stored. Pick a random id in the total length of the quests array and access the quest template from the map?

Reply:
yeah. We've confirmed this new approach to be the way forward.

I usually grab the gas figures by deploying the project to my local ganache, and running the commands from the truffle console, which reports on the gas used. You can also type .estimateGas before the parameters in a call ie
    completeQuest.estimateGas(123)
in case you didn't want to write the action to the chain at that moment, but that estimate is going to be less accurate.
Often I have to account for the first time values are written too, which may cost 20k for each value. For example, first time gaining reputation, first experience, first unclaimed skill etc.

Usually I end up running the action a few (2-3) times to see how the gas changes, and I report the last one, since that will be what most users get assuming they don't claim xp after every fight etc.